### ISSUE TITLE
Android GLES2, error trying to load S3TC textures not present in the APK.

### ISSUE THREAD

**Issue by lawnjelly** [godotengine/godot#28308]
**Godot version:**
Godot 3.1 Stable

**OS/device including version:**
Linux Mint 18.2 Cinnamon 64 bit
Intel Core i5-7500T CPU
Intel HD Graphics 630 GPU
Android 4.4.4 Nexus 7 2012

**Issue description:**
When making Android export, if Import S3TC and Import ETC are selected in project settings->vram compression, only ETC is included in the APK file. However on running on the tablet, it attempts to load non-existent S3TC files, and has loads of errors and doesn't work.

**Steps to reproduce:**
Create blank ES2 project, add a box with a spatial material and add a texture to albedo. Set vram compression to import S3TC and ETC (defaults). Run on android device -> error.

However, it is possible to get it to work by removing S3TC from the import list (i.e. import ETC only), delete the .import folder to force reimport, then re-exporting. This time the android build does not attempt to load non-existent S3TC, and loads the ETC correctly.

_Unfortunately_ the desktop build does not work with only ETC selected, and you get lots of warnings about the S3TC versions not being available, so you have to manually change this and redo the imports when you want to make an android export.

**Minimal reproduction project:**
[s3tc_android_bug.zip](https://github.com/godotengine/godot/files/3104355/s3tc_android_bug.zip)

**ADB Logcat:**

> 04-22 18:40:21.760 8675-8689/com.lawnjelly.s3tc W/GodotView: creating OpenGL ES 2.0 context :
04-22 18:40:21.800 8675-8689/com.lawnjelly.s3tc I/godot: OpenGL ES 2.0 Renderer: NVIDIA Tegra 3
04-22 18:40:21.880 482-498/? I/ActivityManager: Displayed com.lawnjelly.s3tc/org.godotengine.godot.Godot: +1s254ms
04-22 18:40:22.430 8675-8689/com.lawnjelly.s3tc E/godot: **ERROR**: Condition ' !f ' is true. returned: ERR_CANT_OPEN
04-22 18:40:22.430 8675-8689/com.lawnjelly.s3tc E/godot:    At: scene/resources/texture.cpp:493:_load_data() - Condition ' !f ' is true. returned: ERR_CANT_OPEN
04-22 18:40:22.430 8675-8689/com.lawnjelly.s3tc E/godot: **ERROR**: Failed loading resource: res://.import/Board.png-45a35c08fcd69cecff72ce41392a754f.s3tc.stex
04-22 18:40:22.430 8675-8689/com.lawnjelly.s3tc E/godot:    At: core/io/resource_loader.cpp:285:_load() - Method/Function Failed, returning: RES()
04-22 18:40:22.430 8675-8689/com.lawnjelly.s3tc E/godot: **ERROR**: Failed loading resource: res://Board.png
04-22 18:40:22.440 8675-8689/com.lawnjelly.s3tc E/godot:    At: core/io/resource_loader.cpp:285:_load() - Method/Function Failed, returning: RES()
04-22 18:40:22.440 8675-8689/com.lawnjelly.s3tc E/godot: **ERROR**: res://Spatial.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Board.png
04-22 18:40:22.440 8675-8689/com.lawnjelly.s3tc E/godot:    At: scene/resources/resource_format_text.cpp:440:poll() - res://Spatial.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Board.png
04-22 18:40:22.440 8675-8689/com.lawnjelly.s3tc E/godot: **ERROR**: Condition ' err != OK ' is true. returned: RES()
04-22 18:40:22.440 8675-8689/com.lawnjelly.s3tc E/godot:    At: core/io/resource_loader.cpp:208:load() - Condition ' err != OK ' is true. returned: RES()
04-22 18:40:22.440 8675-8689/com.lawnjelly.s3tc E/godot: **ERROR**: Failed loading resource: res://Spatial.tscn

**More Info:**
Having a look in the android exporter, exporter.cpp:

	virtual void get_preset_features(const Ref<EditorExportPreset> &p_preset, List<String> *r_features) {

		String driver = ProjectSettings::get_singleton()->get("rendering/quality/driver/driver_name");
		if (driver == "GLES2") {
			r_features->push_back("etc");
		} else if (driver == "GLES3") {
			r_features->push_back("etc2");
			if (ProjectSettings::get_singleton()->get("rendering/quality/driver/fallback_to_gles2")) {
				r_features->push_back("etc");
			}
		}

This looks like to me in android export GLES2 it is maybe choosing to only export etc (which makes sense as there is no need to export S3TC as well, as EXT is mandated in GLES2). However the code that chooses which texture to load also needs to make the same decision, this is probably where the bug lies.

The specific circumstances which may lead to it are : the android device (Nexus 7) supports S3TC and EXT. However, the android studio emulator does not support S3TC (according to OpenGL caps viewer), and it does not make this attempt to load S3TC, and the bug does not occur on the emulator (it goes straight to load ETC and succeeds).

A example way to prevent this, rather than e.g. relying on only the hardware caps of the device, is to write have e.g. a set of bitflags for which compressed texture formats are present in the export, and not attempt to load ones that are not present. I haven't examined this bit of the code.

e.g.
UNCOMPRESSED = 1
S3TC = 2
ETC = 4

APK has 4
Hardware caps is 7
AND them together to get ones available and supported.

**Comment by ssemakin** [comment]
I think I'm having a similar issue. I also managed to reproduce this issue with a simple project. My setup is this: I create a simple GLES2 project, create a spatial material with the image texture on it, I export to iOS, install on the phone and get the error pasted below. Basically all I see is the black screen. If I change project to GLES3, then it works but there's different types of issues with GLES3 on mobile.

This is the error I see.

```
2019-06-05 11:52:08.526124+0300 export-tester[16002:5966467] OpenGL ES 2.0 Renderer: Apple A10 GPU
OpenGL ES 2.0 Renderer: Apple A10 GPU
2019-06-05 11:52:08.623126+0300 export-tester[16002:5966467] [MC] System group container for systemgroup.com.apple.configurationprofiles path is /private/var/containers/Shared/SystemGroup/systemgroup.com.apple.configurationprofiles
2019-06-05 11:52:08.648576+0300 export-tester[16002:5966467] [MC] Reading from public effective user settings.
2019-06-05 11:52:09.148913+0300 export-tester[16002:5966467] **ERROR**: Failed loading resource: res://icon.png
2019-06-05 11:52:09.148943+0300 export-tester[16002:5966467]    At: core/io/resource_loader.cpp:285:_load() - Method/Function Failed, returning: RES()
**ERROR**: Failed loading resource: res://icon.png
   At: core/io/resource_loader.cpp:285:_load() - Method/Function Failed, returning: RES()
2019-06-05 11:52:09.148993+0300 export-tester[16002:5966467] **ERROR**: res://materials/spatialmaterial_test.tres:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://icon.png
2019-06-05 11:52:09.149007+0300 export-tester[16002:5966467]    At: scene/resources/resource_format_text.cpp:440:poll() - res://materials/spatialmaterial_test.tres:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://icon.png
**ERROR**: res://materials/spatialmaterial_test.tres:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://icon.png
   At: scene/resources/resource_format_text.cpp:440:poll() - res://materials/spatialmaterial_test.tres:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://icon.png
2019-06-05 11:52:09.149027+0300 export-tester[16002:5966467] **ERROR**: Condition ' err != OK ' is true. returned: RES()
2019-06-05 11:52:09.149047+0300 export-tester[16002:5966467]    At: core/io/resource_loader.cpp:208:load() - Condition ' err != OK ' is true. returned: RES()
**ERROR**: Condition ' err != OK ' is true. returned: RES()
   At: core/io/resource_loader.cpp:208:load() - Condition ' err != OK ' is true. returned: RES()
2019-06-05 11:52:09.149277+0300 export-tester[16002:5966467] **ERROR**: Failed loading resource: res://materials/spatialmaterial_test.tres
2019-06-05 11:52:09.149287+0300 export-tester[16002:5966467]    At: core/io/resource_loader.cpp:285:_load() - Method/Function Failed, returning: RES()
**ERROR**: Failed loading resource: res://materials/spatialmaterial_test.tres
   At: core/io/resource_loader.cpp:285:_load() - Method/Function Failed, returning: RES()
2019-06-05 11:52:09.149408+0300 export-tester[16002:5966467] **ERROR**: res://scenes/main_test_scene.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://materials/spatialmaterial_test.tres
2019-06-05 11:52:09.149452+0300 export-tester[16002:
```

**Comment by lawnjelly** [comment]
I have a fix for this, it is quite simple.

The problem occurs because although the android exporter only exports ETC, there may still be a reference to S3TC files in the .import file for the textures (if you are using S3TC on desktop). If the Android device does not support S3TC, it does not attempt to load the file and there is no problem. If however it does support S3TC it attempts to load the file and fails, and exits the game.

There are several ways of solving it:

1. Remove the reference from the .import file
2. Prevent the loader from attempting to load S3TC
3. Pretend there is no hardware support for S3TC on Android devices

I have gone for option 3 as it was easiest for me to implement.

The change is in rasterizer_storage_gles2.cpp, line 5593:

Old version:
```
config.float_texture_supported = config.extensions.has("GL_ARB_texture_float") || config.extensions.has("GL_OES_texture_float");
config.s3tc_supported = config.extensions.has("GL_EXT_texture_compression_s3tc") || config.extensions.has("WEBGL_compressed_texture_s3tc");
config.etc1_supported = config.extensions.has("GL_OES_compressed_ETC1_RGB8_texture") || config.extensions.has("WEBGL_compressed_texture_etc1");
config.pvrtc_supported = config.extensions.has("IMG_texture_compression_pvrtc");
config.support_npot_repeat_mipmap = config.extensions.has("GL_OES_texture_npot");
```

Fixed version:
```
config.float_texture_supported = config.extensions.has("GL_ARB_texture_float") || config.extensions.has("GL_OES_texture_float");
config.s3tc_supported = config.extensions.has("GL_EXT_texture_compression_s3tc") || config.extensions.has("WEBGL_compressed_texture_s3tc");
config.etc1_supported = config.extensions.has("GL_OES_compressed_ETC1_RGB8_texture") || config.extensions.has("WEBGL_compressed_texture_etc1");
config.pvrtc_supported = config.extensions.has("IMG_texture_compression_pvrtc");
config.support_npot_repeat_mipmap = config.extensions.has("GL_OES_texture_npot");

// Special override for running under android.
// The Android exporter (android exporter, exporter.cpp, get_preset_features())
// specifically prevents s3tc being written to the APK, and this makes sense
// as ETC is mandated in GLES 2.0. However the .import file still contains a reference to the s3tc
// file, and any Android device that support s3tc will attempt to load the resource and fail,
// causing exit.
// An simple way to prevent this circumstance is to pretend that hardware support does not exist
// for S3TC on all android devices, which prevents the attempt at loading.
// An alternative method maybe to remove the reference to the s3tc in the .import file.
#if ANDROID_ENABLED
config.s3tc_supported = false;
#endif
```
If any of you guys who use git can put this in (or one of the other alternatives if you prefer) that would be great (I'll try and learn how to use git!!). :smile:

**Comment by lawnjelly** [comment]
> I think I'm having a similar issue. I also managed to reproduce this issue with a simple project. My setup is this: I create a simple GLES2 project, create a spatial material with the image texture on it, I export to iOS, install on the phone and get the error pasted below. Basically all I see is the black screen. If I change project to GLES3, then it works but there's different types of issues with GLES3 on mobile.

It does sound similarish. I don't have any iOS devices so I can't investigate really but can you look inside the exported IPA file and see what versions (if at all) of the png it has in the .import folder? If you can open the logo.png.import file in text editor it should tell you what files it will try to load. I hope this works I have no experience with iOS.

If there is no png file in the .import folder (maybe you would need to reimport the texture as lossless in the Godot IDE?) then perhaps adding ETC textures to your project might get it exported. Sorry I would like to help more.

**Comment by lawnjelly** [comment]
I have spent a bit more time researching this issue and although my above fix addresses the problem in my case, I am now suspecting it may be exposing a wider logic flaw in the whole resource exporting / loading system in Godot (unless the logic is already correct, and this is a simpler bug). To recap, some possible fixes:

_1. Remove the reference from the .import file
2. Prevent the loader from attempting to load S3TC
3. Pretend there is no hardware support for S3TC on Android devices_

I am now tending towards feeling that in order to address this and future bugs of this type with other resources it would be better to implement (1) or (2). Exploring option 1, it appears that the main export happens in this function:

`EditorExportPlatform::export_project_files`

Where it queries the exporter for a feature list of formats that are desired for export (which may not include s3tc for example). It then only exports those that are on the feature list.

However I think it is just copying the .import file word for word into the exported project, containing references to files that have not been exported. At this point perhaps a function could be written to alter the text of the .import file and remove the references to features that are not on the feature list? Godot could also perhaps flag up a warning if you were trying to export a resource with no features supported.

This would be a bit more involved than the quick fix of method 3, but would be a generic solution that would be easier to maintain in the long run, and may prevent other similar bugs occurring in future. I would feel hesitant to make this change myself, it is really something that should be examined by guys familiar with the resource loader, and needs discussion.

Of course anything messing directly with text is prone to more bugs, so saving the feature list with the exported project (is this done already?) and method 2 may be more attractive in practice.

So at this point I hesitate to make a pull request for method 3, as it is a quick fix rather than addressing the underlying issue - unless it would spur more interest. Thoughts anyone?

**Comment by akien-mga** [comment]
Fixed by #32255.
